doc/tpid/SonicTPIDSettingHLD1.md

```diff
@@ -0,0 +1,386 @@
+PORT and LAG objects. This capability query result will be stored in the STATE\_DB so that the front-end CLI handler will be able to check and
+reject TPID setting on port/LAG if such capability is not supported by the SAI vendor running this HW platform. SONiC will be adding two
+entries in STATE\_DB under the existing "SWITCH\_CAPABILITY\|switch" table:
+
+*127.0.0.1:6379\[6\]\> hgetall \"SWITCH\_CAPABILITY\|switch\"*  
+*1) \"**PORT_TPID_CAPABLE**\"*  
+*2) \"**true**\"*  
+*3) \"**LAG_TPID_CAPABLE**\"*  
+*4) \"**false**\"*  
+*5) \"MIRROR\"*  
+*6) \"true\"*  
+*7) \"MIRRORV6\"*  
+*8) \"true\"*  
+*9) \"ACL\_ACTIONS\|INGRESS\"*  
+*10) \"PACKET\_ACTION,MIRROR\_INGRESS\_ACTION\"*  
+*11) \"ACL\_ACTIONS\|EGRESS\"*  
+*12) \"PACKET\_ACTION\"*  
+*13) \"ACL\_ACTION\|PACKET\_ACTION\"*  
+*14) \"DROP,FORWARD\"*  
+*127.0.0.1:6379\[6\]\>*  
+
+For the STATE\_DB shown above this switch/router is only capable of handling PORT TPID setting but not for LAG TPID setting.
+
+**TPID Configuration Flow**
+---------------------------
+
+Port/LAG TPID is configured by either CLI or Restful API talking to CFG manager, which check against State DB to ensure that the corresponding
+object TPID capability is "true" before proceed to update the corresponding port/LAG objects inside CFG DB. portsyncd once received
+the CFG Update notification will then push the config into PORT\_TABLE/LAG\_TABLE inside APP DB. If there is any change with
+PORT\_TABLE/LAG\_TABLE, its subscriber, PortsOrch, will write into ASIC DB and SYNCd will go through SAI to program the ASIC with the configured
+TPID value for the specified Port/LAG object.
+
+The following diagram summarizes how TPID configuration is handled within SONiC:
+
+![TPID Config Flow](TPID_ConfigFlow.png)
+
+
+**CFG Mgr TPID Setting Handling**
+---------------------------------
+
+As mentioned earlier, TPID config is only allowed if SAI capability check passes. Here is a sample CLI config attempt output on the config session
+if the SAI capability is not yet available:
+```
+admin@SONiC:~$ sudo config interface tpid Ethernet64 0x9200
+System not ready to accept TPID config. Please try again later.
+```
+
+Sample CLI config attempt output on the config session if the SAI capability for LAG is "false":
+```
+admin@SONiC:~$ sudo config interface tpid PortChannel0002 0x9200
+HW is not capable to support PortChannel TPID config.
+```
+
+In case a port is already configured as part of a LAG, TPID config directly to the LAG member port is not allowed and will fail as following:
+```
+admin@SONiC:~$ sudo config interface tpid Ethernet4 0x9200
+Ethernet4 is already member of PortChannel0002. Set TPID NOT allowed.
+```
+
+To simplify the handling of TPID setting to avoid complex TPID usage counters and validation checking, we have simplified it by restricting
+the configurable values to the commonly used TPID values: 0x8100 (default), 0x9100, 0x9200, and 0x88A8. If user attempt to configure any

```

Comment:
How does an application know what are the supported TPID's by ASIC?  Is there any SAI support to pull these capabilities? Should it be part of switch capabilities in addition to TPID port & port channel capability?

Reply:
Please refer to the SAI proposal document for details.
you can find it here:
     https://github.com/opencomputeproject/SAI/pull/1089
In summary: The SAI proposal details that the SAI HW vendor when replying to the sai_query_attribute_capability() API that indicate it is capable of supporting the TPID attribute on PORT on or LAG means that it is able to support TPID values of 0x8100, 0x9100, 0x9200, and 0x88A8. 
If the SAI  HW vendor is not able to support these, then it should indicate it is NOT capable to handle this attribute to the Port/LAG object.
Hope this answered your question.

Reply:
sounds good.